contracts/contracts/HMAC_sha512_256.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.0;
+
+import {sha512_256} from "./Sapphire.sol";
+
+// Note that the SHA512_256 block size is 128 bytes, while the output is 32 bytes
+uint256 constant SHA512_256_BLOCK_SIZE = 128;
+
+// We don't (yet) have the MCOPY opcode, so use the IDENTITY precompile
+uint256 constant PRECOMPILE_IDENTITY_ADDRESS = 0x4;
+
+// HMAC block-sized inner padding
+bytes32 constant HMAC_IPAD = 0x3636363636363636363636363636363636363636363636363636363636363636;
+
+// OPAD ^ IPAD, (OPAD = 0x5c)
+bytes32 constant HMAC_OPAD_XOR_IPAD = 0x6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a6a;
+
+/**
+ * @notice Implements HMAC using SHA512-256.
+ * @dev https://en.wikipedia.org/wiki/HMAC
+ * @param key the secret key.
+ * @param message the message to be authenticated.
+ *
+ * #### Example
+ *
+ * ```solidity
+ * bytes memory key = "arbitrary length key";
+ * bytes memory message = "arbitrary length message";
+ * bytes32 hmac = HMAC_sha512_256(key, message)
+ * ```
+ */
+function HMAC_sha512_256(bytes memory key, bytes memory message)
```

Comment:
Was thinking about the function name. Solidity suggests mixed case for functions: https://docs.soliditylang.org/en/latest/style-guide.html#function-names

But since we're already having the underscore `sha512_256` then name it `hmac_sha512_256`? I would avoid upper cases.

```suggestion
function hmac_sha512_256(bytes memory key, bytes memory message)
```

Reply:
I think HMAC_SHA512_256 is a better name, it is more common to use upper case letters to form acronyms.

Reply:
That is a very good question.

Lets create an issue for it, then at the next sprint meeting we can decide if we go ahead with the rename. Then I can submit a PR for it.

Reply:
Some notes for future discussin: renaming a variable of a function which has already been used in code will cause annoying breakages. We should defer the rename to sapphire-contracts 1.x so as to maintain source compatibility